Performing DCR in Comoponent Startup Time Should be Improved

**Description:**
Currently SP tries to do DCR at OSGI component startup time. If the IDP is not up at that time, this will result in logs similar to [1], and also might slow down the server startup time.  in addition, this creates a coupling between SP and IDP as well. This can be improved by attempting the DCR call when a user is logging in. 

[1]. [2019-05-14 05:37:25,700]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.event.simulator.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,701]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.event.simulator.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,701]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.stream.processor.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,702]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.analytics.apim.rest.api.file:1.0.12). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,702]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,703]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,703]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,703]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,703]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.permissions.rest.api:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,704]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,704]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,705]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state until Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.statistics:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:25,705]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state until Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.statistics:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,697]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.event.simulator.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,697]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.event.simulator.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,697]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.stream.processor.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,698]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.analytics.apim.rest.api.file:1.0.12). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,698]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,698]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,698]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,699]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,699]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.permissions.rest.api:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,700]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,700]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,700]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state until Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.statistics:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:37:55,700]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state until Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.statistics:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,696]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.event.simulator.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,697]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.event.simulator.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,697]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component carbon-deployment-service from bundle(org.wso2.carbon.deployment.engine:5.2.0) is in the pending state until Capability org.wso2.carbon.deployment.engine.Deployer from bundle(org.wso2.carbon.stream.processor.core:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,698]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.analytics.apim.rest.api.file:1.0.12). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,698]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,698]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,698]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,699]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.event.simulator.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,699]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.permissions.rest.api:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,699]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,699]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state, because of the Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.core:2.0.423). If you've registered this capability as an OSGi service, you need to declare it using the Carbon-Component manifest header. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,700]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state until Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.statistics:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:25,700]  WARN {org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver} - Startup component wso2-microservices-server from bundle(msf4j-core:2.6.4) is in the pending state until Capability org.wso2.msf4j.Microservice from bundle(org.wso2.carbon.stream.processor.statistics:2.0.423) is available as an OSGi service. Refer the Startup Order Resolver documentation for information.
[2019-05-14 05:38:41,781] ERROR {io.cellery.observability.api.internal.ObservabilityDS} - Error occured while activating the Observability API bundle io.cellery.observability.api.exception.oidc.OIDCProviderException: Error occured while registering client
	at io.cellery.observability.api.auth.OIDCOauthManager.getClientCredentials(OIDCOauthManager.java:114)
	at io.cellery.observability.api.auth.OIDCOauthManager.<init>(OIDCOauthManager.java:74)
	at io.cellery.observability.api.internal.ObservabilityDS.start(ObservabilityDS.java:81)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.internal.ds.model.ServiceComponent.activate(ServiceComponent.java:235)
	at org.eclipse.equinox.internal.ds.model.ServiceComponentProp.activate(ServiceComponentProp.java:146)
	at org.eclipse.equinox.internal.ds.model.ServiceComponentProp.build(ServiceComponentProp.java:345)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponent(InstanceProcess.java:620)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponents(InstanceProcess.java:197)
	at org.eclipse.equinox.internal.ds.Resolver.getEligible(Resolver.java:343)
	at org.eclipse.equinox.internal.ds.SCRManager.serviceChanged(SCRManager.java:222)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:915)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:127)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:225)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:464)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:482)
	at io.cellery.observability.model.generator.internal.ModelServiceComponent.start(ModelServiceComponent.java:45)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.internal.ds.model.ServiceComponent.activate(ServiceComponent.java:235)
	at org.eclipse.equinox.internal.ds.model.ServiceComponentProp.activate(ServiceComponentProp.java:146)
	at org.eclipse.equinox.internal.ds.model.ServiceComponentProp.build(ServiceComponentProp.java:345)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponent(InstanceProcess.java:620)
	at org.eclipse.equinox.internal.ds.InstanceProcess.buildComponents(InstanceProcess.java:197)
	at org.eclipse.equinox.internal.ds.Resolver.getEligible(Resolver.java:343)
	at org.eclipse.equinox.internal.ds.SCRManager.serviceChanged(SCRManager.java:222)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:915)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:127)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:225)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:464)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:482)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:999)
	at org.wso2.carbon.datasource.core.internal.DataSourceListenerComponent.onAllRequiredCapabilitiesAvailable(DataSourceListenerComponent.java:128)
	at org.wso2.carbon.kernel.internal.startupresolver.StartupComponentManager.lambda$notifySatisfiableComponents$7(StartupComponentManager.java:266)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at org.wso2.carbon.kernel.internal.startupresolver.StartupComponentManager.notifySatisfiableComponents(StartupComponentManager.java:252)
	at org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver$1.run(StartupOrderResolver.java:204)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
Caused by: org.apache.http.conn.HttpHostConnectException: Connect to idp.cellery-system:443 [idp.cellery-system/10.19.253.72] failed: Connection timed out (Connection timed out)
	at org.apache.http.impl.conn.HttpClientConnectionOperator.connect(HttpClientConnectionOperator.java:142)
	at org.apache.http.impl.conn.BasicHttpClientConnectionManager.connect(BasicHttpClientConnectionManager.java:318)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:363)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:219)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:195)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:86)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:108)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:106)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:57)
	at io.cellery.observability.api.auth.OIDCOauthManager.getClientCredentials(OIDCOauthManager.java:90)
	... 53 more
Caused by: java.net.ConnectException: Connection timed out (Connection timed out)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:244)
	at org.apache.http.impl.conn.HttpClientConnectionOperator.connect(HttpClientConnectionOperator.java:125)
	... 64 more

[2019-05-14 05:38:41,795] ERROR {org.eclipse.equinox.ds} - [SCR] Exception while activating instance io.cellery.observability.api.internal.ObservabilityDS@38ff233e of component org.wso2.carbon.governance.dependency.model.internal.ObservabilityDS  java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.internal.ds.model.ServiceComponent.activate(ServiceComponent.java:235)

--- a/components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java
+++ b/components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java
@@ -20,6 +20,7 @@ package io.cellery.observability.api;
 
 import io.cellery.observability.api.bean.CelleryConfig;
 import io.cellery.observability.api.exception.APIInvocationException;
+import io.cellery.observability.api.exception.oidc.OIDCProviderException;
 import io.cellery.observability.api.internal.ServiceHolder;
 import org.apache.log4j.Logger;
 import org.apache.oltu.oauth2.client.OAuthClient;
@@ -82,7 +83,7 @@ public class UserAuthenticationAPI {
             NewCookie cookie = new NewCookie(Constants.HTTP_ONLY_SESSION_COOKIE, tokenParts[1],
                     "/", "", "", 3600, false, true);
             return Response.ok().cookie(cookie).entity(responseMap).build();
-        } catch (ConfigurationException | OAuthProblemException | OAuthSystemException e) {
+        } catch (ConfigurationException | OAuthProblemException | OAuthSystemException | OIDCProviderException e) {
             throw new APIInvocationException("Error while getting tokens from Token endpoint", e);
         }
     }

--- a/components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/auth/OIDCOauthManager.java
+++ b/components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/auth/OIDCOauthManager.java
@@ -61,18 +61,25 @@ public class OIDCOauthManager {
     private String clientId;
     private char[] clientSecret;
 
-    public String getClientId() {
-        return clientId;
+    public String getClientId() throws OIDCProviderException {
+        if (this.clientId == null) {
+            synchronized (this) {
+                performDCR();
+            }
+        }
+        return this.clientId;
     }
 
     public String getClientSecret() {
         return String.valueOf(this.clientSecret);
     }
 
-    public OIDCOauthManager() throws OIDCProviderException {
-        JSONObject clientJson = getClientCredentials();
-        this.clientId = clientJson.getString(Constants.CLIENT_ID_TXT);
-        this.clientSecret = clientJson.getString(Constants.CLIENT_SECRET_TXT).toCharArray();
+    public OIDCOauthManager() {
+        try {
+            performDCR();
+        } catch (OIDCProviderException e) {
+            log.warn("DCR failure due to IDP unavailability, will be re-attempted when a user logs in");
+        }
     }
 
     /**
@@ -269,5 +276,16 @@ public class OIDCOauthManager {
             closeResources(inputStreamReader, bufReader);
         }
         return builder;
+    }
+
+    /**
+     * This method will be used to make DCR and set the client credentials obtained from IDP.
+     */
+    private void performDCR() throws OIDCProviderException {
+        if (this.clientId == null) {
+            JSONObject clientJson = getClientCredentials();
+            this.clientId = clientJson.getString(Constants.CLIENT_ID_TXT);
+            this.clientSecret = clientJson.getString(Constants.CLIENT_SECRET_TXT).toCharArray();
+        }
     }
 }

